### ISSUE TITLE
optimize.minimize: Presence of callback causes method TNC to fail 

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#14565]
Ran across this while working on gh-13096: the presence of a callback function can cause TNC to report failure on a problem it otherwise solves correctly.

#### Reproducing code example:

```python3
import numpy as np
from scipy import optimize

np.random.seed(0)
x0 = np.random.rand(4)
lb = [0, 2, -1, -1]
ub = [3, 2, 2, -1]
bounds = list(zip(lb, ub))

def callback(x):
    pass

res = optimize.minimize(optimize.rosen, x0, method="TNC",
                        jac=optimize.rosen_der, bounds=bounds,
                        callback=callback)

## Direct use of `fmin_tnc` has the same issue
# res = optimize.fmin_tnc(optimize.rosen, x0, optimize.rosen_der, bounds=bounds,
#                         callback=callback)
print(res)
```

#### Error message:
```
     fun: 1289.4752548690574
     jac: array([ 4.89571916e-06,  2.49847489e+03,  2.52086570e-07, -3.54799818e+02])
 message: 'Linear search failed'
    nfev: 39
     nit: 5
  status: 4
 success: False
       x: array([ 1.41369616,  2.        ,  0.87977218, -1.        ])

```
On the other hand, if `callback=None`, the function reports success with essentially the same solution but fewer function evaluations and one more iteration.

#### Scipy/Numpy/Python version information:
```
1.8.0.dev0+1528.5288695 1.20.1 sys.version_info(major=3, minor=7, micro=9, releaselevel='final', serial=0)
```
@andyfaff

**Comment by andyfaff** [comment]
Does that happen without any equal bounds?

**Comment by mdhaber** [comment]
Not when I make `lb = [0, 1, -1, -2]`. (But it happens on `master`, and gh-13096 has no effect because analytical gradients are given. If gh-13096 were to remove the variables even when analytical gradients are given, it would fix this issue.)

**Comment by mdhaber** [comment]
Also, in the original example, whether `callback` is provided or not, "TNC" does not respect the bounds on every iteration.

```python3
import numpy as np
from scipy import optimize

np.random.seed(0)
x0 = np.random.rand(4)
lb = np.array([0, 2, -1, -1])
ub = np.array([3, 2, 2, -1])
i_eb = (lb==ub)
bounds = list(zip(lb, ub))

def func(x):
    np.testing.assert_allclose(x[i_eb], lb[i_eb])
    return optimize.rosen(x)

res = optimize.minimize(func, x0, method="TNC",
                        jac=optimize.rosen_der, bounds=bounds)

print(res)
```
gives:
```
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 2 / 2 (100%)
Max absolute difference: 1.54488318
Max relative difference: 1.54488318
 x: array([0.715189, 0.544883])
 y: array([ 2, -1])
```
This would also be fixed if gh-13096 were to remove the fixed variables altogether even when analytical gradients are given.

Should this be filed as a separate issue?

**Comment by andyfaff** [comment]
This may have been reported before, but I can't remember.

We can add TNC to #13096. Perhaps we should remove the variables for TNC even if the analytic Jacobian is provided. I'd be reluctant to do that for SLSQP and LBFGSB because they're currently known to work.

**Comment by mdhaber** [comment]
Yes, I had TNC in `test_equal_bounds2`; that's how this came up.

I'm now noticing that `slsqp` can call `constraints` with the provided `x0`, and some of its elements can be outside of bounds (whether or not there are some fixed variables). Is this supposed to happen, or should the initial guess be clipped to be within the bounds?

**Comment by andyfaff** [comment]
The constraint jacobian has its [values clipped](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L295), to prevent `approx_derivative` complaining, but we're not currently clipping x supplied to the constraint function itself. That should probably happen in [_eval_constraint](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L465).

The initial guess is clipped to the bounds to [start with](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L262).

I find these Fortran codes very annoying, it's sloppy of it to generate `x` that lie outside the bounds.

**Comment by andyfaff** [comment]
This issue should be fixed by #14882. The TNC c-code mutates x when it does the callback.

**Comment by andyfaff** [comment]
Closed by #14882
